dips/dip-11.md

```diff
@@ -6,3 +6,339 @@ status: Draft
+
+`WriteSet` transactions use a special `writeset_epilogue` function from the `DiemAccount` module. The `writeset_epilogue` calls the standard epilogue to increment the `sequence_number`, emits an `AdminTransactionEvent`, and if the `WriteSetPayload` is a `Direct` value, it also emits a `NewEpochEvent` to trigger reconfiguration. For a `Script` value in the `WriteSetPayload`, it is the responsibility of the code in the script to determine whether a reconfiguration is necessary, and if so, to emit the appropriate `NewEpochEvent`. If the epilogue does not execute successfully, the status code is set to `UNEXPECTED_ERROR_FROM_KNOWN_MOVE_FUNCTION`.
+
+## Account events
+
+Whenever the value of one of the account’s balances is changed, specific events are emitted that describe the state change that occurred.
+
+Whenever an account’s balance resource receives funds a `ReceivedPaymentEvent` event with the following structure is recorded on the `received_events` event handle in the account’s `DiemAccount` resource. 
+
+```
+struct ReceivedPaymentEvent {
+    amount: u64,
+    currency_code: vector<u8>,
+    payer: address,
+    metadata: vector<u8>,
+}
+```
+
+The `amount` field of the recorded event specifies the  amount by which the account’s balance was increased, and the `currency_code` specifies the balance of the currency in which the funds were added. The `payer` field specifies the address from which the deposited funds were withdrawn, and the `metadata` field contains any metadata attached to funds by the `payer`.
+
+When an account’s balance resource has funds withdrawn from it a `SentPaymentEvent` event with the following structure is recorded on the `sent_events` event handle in the debited account’s `DiemAccount` resource. 
+
+```
+struct SentPaymentEvent {
+    amount: u64,
+    currency_code: vector<u8>,
+    payee: address,
+    metadata: vector<u8>,
+}
+```
+
+The `amount` and `currency_code` fields of the recorded event specify the amount and the currency of the withdrawal from the account’s balance. The `payee` field specifies where the withdrawn funds are re-deposited, and the `metadata` fields contains any metadata that the `payee` wishes to attach to the payment.
+
+### Account Event Invariants
+
+The `SentPaymentEvent` and  `ReceivedPaymentEvent`, along with gas payments for a particular account, can be used to derive the account’s balance. To see this, let H be the event handle that received one of the above `SentPaymentEvent` or `ReceivedPaymentEvent` events, and let A be the account address under which this event handle is published. The following invariants hold:
+
+* The `currency_code` balance under A increases by `amount` if and only if a `ReceivedPaymentEvent` `{amount, currency_code, payer}` is recorded on handle H.
+* If a `SentPaymentEvent` `{amount, currency_code, payee}` is recorded on handle H, A’s `currency_code` balance decreases by the specified `amount`.
```

Comment:
It would not be correct to specify that here because a single transaction can theoretically contain multiple `SentPaymentEvent`'s (e.g., if we expose a transaction type that pays multiple users), but will only have a single gas charge. Thus, I think having a separate bullet about gas charges makes sense.

Reply:
Got it.  These invariants are held with respect to the events not the transaction that generated these events.